pkg/deploy/controller/deployment/controller.go

```diff
@@ -278,67 +345,31 @@ func (c *DeploymentController) cleanupDeployerPods(deployment *kapi.ReplicationC
-	if config, _ := c.decodeConfig(deployment); config != nil {
+	if config, _ := deployutil.DecodeDeploymentConfig(deployment, c.codec); config != nil {
 		c.recorder.Eventf(config, eventType, title, fmt.Sprintf("%s: %s", deployment.Name, message))
 	} else {
 		c.recorder.Eventf(deployment, eventType, title, message)
 	}
 }
 
-// deploymentClient abstracts access to deployments.
-type deploymentClient interface {
-	getDeployment(namespace, name string) (*kapi.ReplicationController, error)
-	updateDeployment(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error)
-}
-
-// podClient abstracts access to pods.
-type podClient interface {
-	getPod(namespace, name string) (*kapi.Pod, error)
-	createPod(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	deletePod(namespace, name string) error
-	updatePod(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	getDeployerPodsFor(namespace, name string) ([]kapi.Pod, error)
-}
-
-// deploymentClientImpl is a pluggable deploymentClient.
-type deploymentClientImpl struct {
-	getDeploymentFunc    func(namespace, name string) (*kapi.ReplicationController, error)
-	updateDeploymentFunc func(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error)
-}
-
-func (i *deploymentClientImpl) getDeployment(namespace, name string) (*kapi.ReplicationController, error) {
-	return i.getDeploymentFunc(namespace, name)
-}
-
-func (i *deploymentClientImpl) updateDeployment(namespace string, deployment *kapi.ReplicationController) (*kapi.ReplicationController, error) {
-	return i.updateDeploymentFunc(namespace, deployment)
-}
-
-// podClientImpl is a pluggable podClient.
-type podClientImpl struct {
-	getPodFunc             func(namespace, name string) (*kapi.Pod, error)
-	createPodFunc          func(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	deletePodFunc          func(namespace, name string) error
-	updatePodFunc          func(namespace string, pod *kapi.Pod) (*kapi.Pod, error)
-	getDeployerPodsForFunc func(namespace, name string) ([]kapi.Pod, error)
-}
-
-func (i *podClientImpl) getPod(namespace, name string) (*kapi.Pod, error) {
-	return i.getPodFunc(namespace, name)
-}
-
-func (i *podClientImpl) createPod(namespace string, pod *kapi.Pod) (*kapi.Pod, error) {
-	return i.createPodFunc(namespace, pod)
-}
+func (c *DeploymentController) handleErr(err error, key interface{}, deployment *kapi.ReplicationController) {
+	if err == nil {
+		return
+	}
 
-func (i *podClientImpl) deletePod(namespace, name string) error {
-	return i.deletePodFunc(namespace, name)
-}
+	if _, isFatal := err.(fatalError); isFatal {
+		utilruntime.HandleError(err)
+		c.queue.Forget(key)
+		return
+	}
 
-func (i *podClientImpl) updatePod(namespace string, pod *kapi.Pod) (*kapi.Pod, error) {
-	return i.updatePodFunc(namespace, pod)
-}
+	if c.queue.NumRequeues(key) < 2 {
```

Comment:
Once we collapse the deployer pod controller, we can start retrying updates on the spot.